File: src/131PalindromePartitioning/131PalindromePartitioning.go
package main

import (
	"fmt"
)

func isPalindrome(s string) bool {
	lo := 0
	hi := len(s) - 1
	for lo <= hi {
		if s[lo] != s[hi] {
			return false
		} else {
			lo++
			hi--
		}
	}
	return true
}
func backtracking(s string, results [][]string, result []string) [][]string {
	if s == "" {
		r := make([]string, len(result))
		copy(r, result)
		results = append(results, r)
	} else {
		for k := 1; k <= len(s); k++ {
			if isPalindrome(s[:k]) {
				result = append(result, s[:k])
				results = backtracking(s[k:], results, result)
				result = result[:len(result)-1]
			}
		}
	}
	return results
}

func partition(s string) [][]string {
	var results [][]string
	var result []string
	results = backtracking(s, results, result)
	return results
}

func main() {
	fmt.Printf("%v\n", partition("aabcbaa"))
}

File: src/131PalindromePartitioning/131PalindromePartitioning.cpp
class Solution {
public:
    vector<vector<string>> partition(string s) {
        return partition(s, 0, int(s.size())-1);
    }
    
    vector<vector<string>> partition(string& s, int lo, int hi) {
        vector<vector<string>> results;
        for(int i=lo; i<=hi; ++i){
            if(is_palindrome(s, lo, i)){
                auto ss = s.substr(lo,i-lo+1);
                auto result = partition(s, i+1, hi);
                if(!result.empty()){
                    for(auto& res : result){
                        res.insert(res.begin(), ss);
                    }
                }else{
                    result.push_back(vector<string>{ss});
                }
                
                
                copy(result.begin(), result.end(), back_inserter(results));
            }    
        }    
        return results;
    }
    
    bool is_palindrome(string& s, int lo, int hi){
        while(lo<hi){
            if(s[lo++]!=s[hi--]){
                return false;
            }
        }
        return true;
    }
};
